refc!: express-slow-down v2

nfriedly opened the pull request:
A few major changes here:

* [x] Convert express-slow-down to use express-rate-limit internally
  * This should get us automatic feature parity with erl, including store compatibility (see #41).
* [x] Allow `delayMs` to be a function (see #21)
  * passed `used`,  `req`, and `res`; called with `options` as `this`
  * [x] todo: consider changing the behavior when it's a number to just be a fixed delay (breaking change, but probably more straightforward)
* [x] removed `headers` option
* [x] Simplify after https://github.com/express-rate-limit/express-rate-limit/discussions/369 is implemented
* [x] Wait for a way to disable the max=0 warning
  * PR: https://github.com/express-rate-limit/express-rate-limit/pull/392 
* [x] typescript
* [x] ESM?
* [x] Update test matrix to stay on current and latest LTS release for win/lin/mac, + node 14 on lin
  * Should sidestep timeout issues with node 14 on mac
  * Also updated actions to latest versions to resolve warnings 
* [x] Warn about changed behavior when `delayMs` is a number rather than a function.

Not included here:
* figure out how to share a store with erl so that requests don't have to be counted twice with separate keys 
  * not likely to be a breaking change
  * should happen in express-rate-limit, not here

Fixes #9 
Fixes #21 
Fixes #41 
Closes #17

gamemaker1 commented:
Hi,

I'm working on the typescript-esm part right now, and I had a couple of questions -

1. why is `validate` disabled by default for the delayer?
2. should we jump straight to node >= 16 support for this package?
   - should we switch to only named export here? or keep both like erl?
3. should I rewrite the tests from scratch? I think a lot of the tests are duplicates of erl tests, we can skip those and just test the delay mechanism.

Also, great work so far :)

nfriedly commented:
> I'm working on the typescript-esm part right now

Cool! Maybe branch off of v2, and then make a PR targeting this v2 branch?

> 1. why is validate disabled by default for the delayer?

Mainly because `max=0` is a valid option here.  I was thinking it might be time to make that more granular, but haven't done it yet.

> 2. should we jump straight to node >= 16 support for this package?
>     *   should we switch to only named export here? or keep both like erl?

Maybe, let me think about it a bit more.

> 3. should I rewrite the tests from scratch? I think a lot of the tests are duplicates of erl tests, we can skip those and just test the delay mechanism.
No, I don't think so

gamemaker1 commented:
> Cool! Maybe branch off of v2, and then make a PR targeting this v2 branch?

Sure!

> Mainly because max=0 is a valid option here. I was thinking it might be time to make that more granular, but haven't done it yet.

Ohkay

> 2 & 3

👍🏾

nfriedly commented:
Ok, I think I have the behavior changes I wanted in place. I started working on a TypeScript conversion last night, and I believe I have the library done (it doesn't have any red squiggelies any more). Still need to do the tests and set up npm run scripts.

gamemaker1 commented:
@nfriedly did you forget to commit the typescript files in the last commit?

nfriedly commented:
🤦‍♂️

nfriedly commented:
As I was working on this I remembered that we had already added support for delayMs being a function, we just never documented it. The v1.6 version is a synchronous function that is passed `req`, and `res`.  I really like the newer design of passing in `used` as the first argument, but I just realized that it's technically already there in `req.slowDown.used` in the old design. So now I need to decide if it's worth the breaking change to have a better function signature. I don't think the breaking change would affect that many people, but still.

Also, in the default function, I'm using the `delayAfter` value via `req.slowDown.limit`, so perhaps either `delayAfter` should be the new second argument, *OR* maybe the `req.slowDown`  object should be the first argument. Destructuring would make it pretty easy to create a function that takes just `used`, e.g. `({used}) => used * 1000`, or `used` and `limit`, e.g. `({used, limit}) => (used - limit) * 1000`.

I want to think about it a bit more, but I'm also interested in hearing other people's thoughts.

nfriedly commented:
I think this is finally done! I'm sorry it's such a monster PR; it wasn't so bad until I did the typescript conversion, but that meant that basically everything got touched. Do you want me to try to turn it into a series of PRs for easier reviewing @gamemaker1 ?

gamemaker1 commented:
Hooray!! Great work, @nfriedly!

Don't worry about it being a monster PR, that was bound to happen with the typescript conversion. I'll work my way through it over the weekend.

nfriedly commented on package.json:
```diff
@@ -60,8 +60,10 @@
+    "lint:code": "xo",
+    "lint:rest": "prettier --check .",
+    "format:code": "xo --fix",
+    "format:rest": "prettier --write .",
```
I had it set up like this initially, and it seemed to cause some thrashing in the .ts files, so I think it's really formatting everything, and not using the exact same rules as xo.

gamemaker1 replied:
I think the problem was the mismatch between the version of XO, Typescript and Prettier. Should be fixed with d451e6c.

nfriedly replied:
Aah, that makes sense.

nfriedly commented on .editorconfig:
```diff
@@ -1,16 +1,20 @@
 
-# Matches multiple files with brace expansion notation
-# Set default charset
-[*.{js},{package.json,.travis.yml}]
-indent_style = space
-indent_size = 2
```
I think we want this for `package.json`, at least

gamemaker1 replied:
Done!

nfriedly commented:
I went back and forth on this in my head, but I think we should add a validation check that warns about the changed behavior when `delayMs` is set to a number. That's basically going to hit everyone who upgrades, but I think the change is significant enough that we need to do it.

I think it could include a couple of options in the error message, e.g. if `delayAfter` is 1 and `delayMs` is 100, it could recommend setting `delayMs` to `(hits) => (hits - 1) * 100` for the old behavior or `() => 100` for the new behavior. And then the warning wouldn't trigger if it was a function.

I'm not sure if we should make express-rate-limit's validator extensible, or just do a standalone one. Standalone would probably be easier.

gamemaker1 commented:
> I went back and forth on this in my head, but I think we should add a validation check that warns about the changed behavior when delayMs is set to a number. That's basically going to hit everyone who upgrades, but I think the change is significant enough that we need to do it.

I agree, we should definitely put in a validation check for this.

> I think it could include a couple of options in the error message, e.g. if delayAfter is 1 and delayMs is 100, it could recommend setting delayMs to (hits) => (hits - 1) * 100 for the old behavior or () => 100 for the new behavior. And then the warning wouldn't trigger if it was a function.

I think the error message would get a bit too long with this; what if we put this in the wiki page instead?

> I'm not sure if we should make express-rate-limit's validator extensible, or just do a standalone one. Standalone would probably be easier.

Standalone sounds good for now.

gamemaker1 commented:
Couldn't fix the flaky test, so I switched the realtime (renamed the file to `integration-test.ts`) tests to run only on linux, similar to the external tests on `express-rate-limit`.

gamemaker1 commented:
Added the validation check, its redirect and the explanation in the wiki page too.

I think the PR is ready to go! Sorry for taking so long with the review 😅

nfriedly commented on source/slow-down.ts:
```diff
@@ -0,0 +1,160 @@
+	if (typeof notUndefinedOptions.delayMs === 'number') {
+		const url = `https://express-rate-limit.github.io/WRN_ESD_DELAYMS/`
+		console.warn(
+			`The behaviour of the 'delayMs' option was changed in express-slow-down v2. See ${url} for more information.`,
```
Oh, right forgot about this. Thank you!

Two thoughts:

1) I think we should wrap it in `new Error(...)` to provide a stack trace.

2) Since this is going to hit *literally everyone*, I think we should provide the fix right in the error message:

```suggestion
			new Error(`The behaviour of the 'delayMs' option was changed in express-slow-down v2:

* For the old behavior, change the delayMs option to
  delayMs: (used, req) => {
	  const delayAfter = req.${notUndefinedOptions.requestPropertyName ?? 'slowDown'}.limit;
	  return (used - delayAfter) * ${notUndefinedOptions.delayMs};
  },

* For the new behavior, change the delayMs option to:
delayMs: () => ${notUndefinedOptions.delayMs},

See ${url} for more information.`),
```


3) It'd be nice if the warning could be disabled. Perhaps we can add a new `options.validate.delayMs` that we accept in ESD, but remove before passing along to ERL? And then maybe change the new behavior bullet point to point towards disabling the check rather than setting a function?

gamemaker1 replied:
Sorry for the late reply, I had an exam today and was preparing for it.

I changed the message to an error, so the stack trace shows up now. I also changed the message to what you suggested, and added the following at the end:

```
Set 'options.validate.delayMs: false' to disable this error message.
```

I added a way to disable the warning, and a test for it too.

gamemaker1 commented:
Let me know if there's anything else, otherwise this pr looks good to go!

nfriedly commented:
Sorry, I've had a lot going on past few days. I will try and get a look at this sometime soon, hopefully today

gamemaker1 commented:
No problem, take your time!

nfriedly commented:
I made a couple of tweaks to the formatting of the message, and then prevented the unknown validation warning from ERL, added a test, and simplified things a bit

gamemaker1 reviewed (approved):
Merge this whenever you're ready!

nfriedly commented:
I think I'm going to wait until next week, because I'd like to give it one last pass, and I'm not going to have time before I leave today

nfriedly commented on package.json:
```diff
@@ -31,20 +27,114 @@
+    "express": ">= 4"
+  },
+  "dependencies": {
+    "express-rate-limit": "7.0.1"
```
We should loosen this up so that someone who wants to use both express-slow-down and express-rate-limit doesn't end up needing two installs of express-rate-limit.

Probably something like this:

```suggestion
    "express-rate-limit": ">= 7"
```

nfriedly replied:
Or perhaps this if we don't want to risk express-rate-limit version 8+ introducing an incompatible change:

```suggestion
    "express-rate-limit": "^7.0.0"
```

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,278 @@
+provide more consistency.
+
+express-slow-down uses
+[express-rate-limit's stores](https://github.com/express-rate-limit/express-rate-limit#store)
```
Op, we'll want to update this after https://github.com/express-rate-limit/express-rate-limit/pull/411 goes live

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,278 @@
+	// ...
+	delayAfter: 1,
+	delayMs: (hits) => hits * 1000,
+	maxDelayMs: 4_000,
```
Not sure I like the underscore here... did you do that, or was it an eslint rule?

```suggestion
	maxDelayMs: 4000,
```

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,278 @@
+[`express-rate-limit`](https://github.com/express-rate-limit/express-rate-limit)
+is used internally, additional options that it supports may be passed in. Some
+of them are listed below; see `express-rate-limit`'s
+[readme](https://github.com/express-rate-limit/express-rate-limit#configuration)
```
We'll also want to update this URL after https://github.com/express-rate-limit/express-rate-limit/pull/411

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,278 @@
+> **Note**: The `limit` (`max`) option is not supported (use `delayAfter`
+> instead), nor are `handler` or the various headers options.
+
+- [`requestPropertyName`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#requestpropertyname)
+- [`skipFailedRequests`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#skipfailedrequests)
+- [`skipSuccessfulRequests`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#skipsuccessfulrequests)
+- [`keyGenerator`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#keygenerator)
+- [`skip`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#skip)
+- [`requestWasSuccessful`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#requestwassuccessful)
+- [`validate`](https://github.com/express-rate-limit/express-rate-limit/blob/main/readme.md#validate)
+- [`store`](https://github.com/express-rate-limit/express-rate-limit#store)
```
...and all of these

nfriedly commented on source/slow-down.ts:
```diff
@@ -0,0 +1,202 @@
+const omitUndefinedOptions = (
+	passedOptions: Partial<Options>,
+): Partial<Options> => {
+	const omittedOptions: Partial<Options> = {}
```
`filteredOptions` might be a better name for this variable; `omittedOptions` suggests to me that it's the opposite of what it actually is.

nfriedly commented on source/slow-down.ts:
```diff
@@ -0,0 +1,202 @@
+		notUndefinedOptions.legacyHeaders ||
+		notUndefinedOptions.standardHeaders
+	)
+		throw new Error(
+			'The headers options were removed in express-slow-down v2.0.0.',
+		)
+
+	if (
+		notUndefinedOptions.max !== undefined ||
+		notUndefinedOptions.limit !== undefined
+	)
+		throw new Error(
+			'The limit/max option is not supported by express-slow-down, please use delayAfter instead.',
+		)
```
(we could also change them to validation checks that log rather than throw)

gamemaker1 replied:
Do you want to move `express-slow-down`'s docs to mintlify too?

nfriedly commented on source/slow-down.ts:
```diff
@@ -0,0 +1,202 @@
+	return omittedOptions
+}
+
+// Todo: consider exporting then extending express-rate-limit's ValidationError
```
Or, maybe we should make the validator a standalone library that both ERL and ESD use? Not in this PR, though.

gamemaker1 replied:
Agreed.

gamemaker1 commented on source/index.ts:
```diff
@@ -0,0 +1,8 @@
+export * from './types.js'
+
+// Export the `slowDown` function as well.
+export { default, default as slowDown } from './slow-down.js'
```
Turns out `export *` does not re-export a default, so I'll keep it as is
